File: Chapter17/src/javaexamples/SearchRequest.java
package javaexamples;

public class SearchRequest {

    private String queryTerm;
    private String[] categories;
    private int resultCount;

    private SearchRequest() { }

    private SearchRequest(String queryTerm, String[] categories, int resultCount) {
        this.queryTerm = queryTerm;
        this.categories = categories;
        this.resultCount = resultCount;
    }

    public String getQueryTerm() {
        return queryTerm;
    }

    public String[] getCategories() {
        return categories;
    }

    public int getResultCount() {
        return resultCount;
    }

    public static class Builder {

        private String queryTerm;
        private String[] categories;
        private int resultCount;

        public Builder queryTerm(String queryTerm) {
            this.queryTerm = queryTerm;
            return this;
        }

        public Builder categories(String[] categories) {
            this.categories = categories;
            return this;
        }

        public Builder resultCount(int resultCount) {
            this.resultCount = resultCount;
            return this;
        }

        public SearchRequest build() {
            return new SearchRequest(queryTerm, categories, resultCount);
        }
    }
}

File: Chapter17/src/kotlinexamples/SearchRequest.kt
package kotlinexamples

data class SearchRequest(
    val queryTerm: String = "",
    val categories: List<String> = listOf(),
    val resultCount: Int = 10
)

fun main() {
    val request1 = SearchRequest()
    val request2 = SearchRequest(
        categories = listOf("Kotlin", "Android"),
        queryTerm = "Kotlin"
    )
    val request = SearchRequest(resultCount = 20)
}